core/src/main/java/io/questdb/griffin/engine/table/AsyncGroupByAtom.java

```diff
@@ -0,0 +1,153 @@
+ *  Copyright (c) 2014-2019 Appsicle
+ *  Copyright (c) 2019-2023 QuestDB
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ ******************************************************************************/
+
+package io.questdb.griffin.engine.table;
+
+import io.questdb.cairo.*;
+import io.questdb.cairo.sql.Function;
+import io.questdb.cairo.sql.SqlExecutionCircuitBreaker;
+import io.questdb.cairo.sql.StatefulAtom;
+import io.questdb.cairo.sql.SymbolTableSource;
+import io.questdb.griffin.PlanSink;
+import io.questdb.griffin.Plannable;
+import io.questdb.griffin.SqlException;
+import io.questdb.griffin.SqlExecutionContext;
+import io.questdb.griffin.engine.PerWorkerLocks;
+import io.questdb.griffin.engine.functions.GroupByFunction;
+import io.questdb.griffin.engine.groupby.GroupByFunctionsUpdater;
+import io.questdb.griffin.engine.groupby.GroupByFunctionsUpdaterFactory;
+import io.questdb.std.BytecodeAssembler;
+import io.questdb.std.Misc;
+import io.questdb.std.ObjList;
+import io.questdb.std.Transient;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.io.Closeable;
+
+public class AsyncGroupByAtom implements StatefulAtom, Closeable, Plannable {
+
+    private final Function filter;
+    private final GroupByFunctionsUpdater functionUpdater;
+    private final RecordSink mapSink;
+    private final ObjList<Function> perWorkerFilters;
+    private final PerWorkerLocks perWorkerLocks;
+
+    public AsyncGroupByAtom(
+            @Transient @NotNull BytecodeAssembler asm,
+            @NotNull CairoConfiguration configuration,
+            @Transient @NotNull ColumnTypes columnTypes,
+            @Transient @NotNull ListColumnFilter listColumnFilter,
+            @NotNull ObjList<GroupByFunction> groupByFunctions,
+            @Nullable Function filter,
+            @Nullable ObjList<Function> perWorkerFilters,
+            int workerCount
+    ) {
+        assert perWorkerFilters == null || perWorkerFilters.size() == workerCount;
+        this.filter = filter;
+        this.perWorkerFilters = perWorkerFilters;
+        functionUpdater = GroupByFunctionsUpdaterFactory.getInstance(asm, groupByFunctions);
+        if (perWorkerFilters != null) {
+            perWorkerLocks = new PerWorkerLocks(configuration, workerCount);
+        } else {
+            perWorkerLocks = null;
+        }
+        mapSink = RecordSinkFactory.getInstance(asm, columnTypes, listColumnFilter, false);
```

Comment:
perhaps worth adding finally with `close()`

Reply:
Done in bb56cdca1b0d85618989ec34046cb246b8111c7f